parser/internal/names/types.go

```diff
@@ -15,6 +15,61 @@ import (
 // otherwise needs to parse the package source to find it.
 type TrackedPackages map[string]string
 
+// Application presents all resolved names throughout the Application
+type Application map[*est.Package]*Resolution
+
+// PkgObjRef resolves the given ast.Node to a package path, object name and type arguments. In the case it fails to
+// resolve, it will return ("", "", nil).
+//
+// If the given node is a index or index list, then the type arguments will be extracted.
+//
+// If the given node is:
+// - an ident, then it is resolved to a package level deceleration in the same file as the ident.
+// - a selector, then it is resolved to an exported ident from the selected package.
+// - any other node will result in nothing being returned
+//
+// Note:
+// - This function accounts for scoping and variable shadowing. (i.e. it will not mistake a variable "foo" in a function
+//   for the package imported as "foo").
+// - This function does not validate the existence of the objName for objects referenced in other packages. As such it
+//   can return references for objects in Go Modules.
+func (a Application) PkgObjRef(inFile *est.File, node ast.Node) (pkgPath string, objName string, typeArguments []ast.Expr) {
+	f, ok := a[inFile.Pkg].Files[inFile]
+	if !ok {
+		return "", "", nil
+	}
+
+	// Resolve the type arguments
+	var typArgs []ast.Expr
+	switch typNode := node.(type) {
+	case *ast.IndexExpr:
+		node = typNode.X
+		typArgs = []ast.Expr{typNode.Index}
```

Comment:
I think we should not blindly unwrap index list expressions here; it becomes imprecise when we don't know whether something accepts type parameters. It would accept code like `var blah = cron.NewJob[int]()` even though `cron.NewJob` isn't generic.

It would also say that `otherpkg.Blah[0]` is a package object expression, but it's not. `otherpkg.Blah` is a package object.

If we want to do unwrapping of type parameters we should require the caller to decide whether type parameters are permissible in a particular location, and then we should arguably parse those type arguments as well. That feels like it belongs in the `parser` package rather than the `names` package though so I'd prefer if we keep the "unwrap index expressions" code outside of the package object resolution

Reply:
I've updated the docs and return names to say `indexArgs` here.

For name resolution, I think this is still valid to have, as it's operating over all names within the application.

> It would also say that otherpkg.Blah[0] is a package object expression

This function say it's a reference to a package object - which it is with index args being type args _or_ an actual index into a map or slice of a package object.